crates/net/network/src/transactions/mod.rs

```diff
@@ -1082,85 +1083,113 @@ where
+                                    // ignore it, otherwise we penalize the peer that sent the bad
+                                    // transaction with the assumption that the peer should have
+                                    // known that this transaction is bad. (e.g. consensus
+                                    // rules)
+                                    if err.is_bad_transaction() && !this.network.is_syncing() {
+                                        debug!(target: "net::tx", ?err, "bad pool transaction import");
+                                        this.on_bad_import(err.hash);
+                                        continue
+                                    }
+                                    this.on_good_import(err.hash);
+                                }
+                            }
+                        }
                     }
                     Err(err) => {
-                        // if we're _currently_ syncing and the transaction is bad we ignore it,
-                        // otherwise we penalize the peer that sent the bad
-                        // transaction with the assumption that the peer should have
-                        // known that this transaction is bad. (e.g. consensus
-                        // rules)
-                        if err.is_bad_transaction() && !this.network.is_syncing() {
-                            debug!(target: "net::tx", ?err, "bad pool transaction import");
-                            this.on_bad_import(err.hash);
-                            continue
-                        }
-                        this.on_good_import(err.hash);
+                        debug!(target: "net::tx", ?err, "bad pool transaction batch import");
                     }
                 }
+
+                some_ready = true;
             }
-        }
 
-        // handle and propagate new transactions
-        let mut new_txs = Vec::new();
-        while let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
-            new_txs.push(hash);
-        }
-        if !new_txs.is_empty() {
-            this.on_new_transactions(new_txs);
-        }
+            // handle and propagate new transactions.
+            //
+            // higher priority! stream is drained
+            //
+            let mut new_txs = Vec::new();
+            while let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
+                new_txs.push(hash);
+                some_ready = true;
+            }
+            if !new_txs.is_empty() {
+                this.on_new_transactions(new_txs);
+            }
 
-        // all channels are fully drained and import futures pending
+            // all channels are fully drained and import futures pending
+            if !some_ready {
+                return Poll::Pending
+            }
 
-        Poll::Pending
+            budget -= 1;
+            if budget <= 0 {
```

Comment:
```suggestion
            if budget == 0 {
```

Reply:
it is ill-advised to make an arithmetic condition like `if x == 0`. if the machine messes up arithmetic a bit, which can always happen, then the loop keeps going.

Reply:
I don't see how this value can be lower than 0

Reply:
yes, the compiler optimises away, thanks for sharing this in chat (incase someone else is reading here later and is curious): https://godbolt.org/z/do6x1Erd6

Reply:
because it's an `i32` by default

Reply:
if we use a `u32` we should at best use `saturating_sub` for same reason: machine might be super hot from churn,  mess up an arithmetic, and then we don't want it to get blocked in an `abs(i32::MIN)` long loop or in the case of `u32`, panic. we want the machine to try and gracefully shutdown when it notices it's past its limits.

Reply:
my bad I expected the budget to be unsigned as the other usages of this pattern, which does make more sense to me